Gems/Atom/Tools/AtomToolsFramework/Code/Include/AtomToolsFramework/Graph/GraphUtil.h

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#pragma once
+
+#include <AzCore/std/sort.h>
+#include <AzCore/std/tuple.h>
+#include <GraphModel/Model/Connection.h>
+#include <GraphModel/Model/Node.h>
+
+namespace AtomToolsFramework
+{
+    // Sort a container of nodes by depth, considering the number and state of input and output slots and connections, for consistent
+    // display and execution order. The function is templatized to account for different container types and constness. 
+    template<typename NodeContainer>
+    void SortNodesInExecutionOrder(NodeContainer& nodes)
```

Comment:
I know we do this all over the codebase, but this actually exceeds my bar for "too big to be in an inlined template function." Why not have the `NodeValueType` here inherit from a base class providing `HasInputSlots`, `HasOutputSlots`, and `GetMaxInputDepth`?

Reply:
The node class is the same regardless of the template type. The template was previously needed because the container type and constness would change. I will see if there's another way to organize it to shrink the template function.